### PR TITLE
tear off via drag

### PR DESCRIPTION
**Pull request by stewart-adaptive**
Update react-redux version to use new hooks
Added DragWrapper to TearOff with drag start/end functionality
Added drag coordinates to layout state/reducer & click coordinates when clicking popout icon.
Record coordinates on drag and open new windows at that position rather than default to center (OpenFin only)

Ive changed Blotter to a functional component as originally I had added the drag code to the component rather than tear off, Ive left the change in as we want to move to FC with hooks anyway, I can remove this change if you want.

TearOff

### CODE REVIEW

**Comment by pmilla1606** [comment]
Looks good to me. Working great locally too.

👍

File: src/client/src/shell/layouts/reducer.ts [hunk 1]
```diff
@@ -27,20 +37,32 @@ export const layoutReducer = (
         case externalWindowDefault.blotterRegion.config.name:
           return {
             ...state,
-            displayBlotter: action.payload.display,
+            blotter: {
```
Comment (bhavesh-desai-scratch) [inline src/client/src/shell/layouts/reducer.ts]: collapse these

File: src/client/src/shell/routes/ShellRoute.tsx [hunk 2]
```diff
@@ -12,18 +12,31 @@ import ReconnectModal from '../components/reconnect-modal'
 import DefaultLayout from '../layouts/DefaultLayout'
 import { BlotterWrapper, AnalyticsWrapper, WorkspaceWrapper, OverflowScroll } from './styled'
 import { GlobalState } from '../../StoreTypes'
-import { displayAnalyticsSelector, displayBlotterSelector } from '../layouts/selectors'
+import { analyticsSelector, blotterSelector } from '../layouts/selectors'
+import { TileLayoutState } from '../layouts'
+import { ExternalWindow } from './../../rt-components/platform/externalWindowDefault'
```
Comment (bhavesh-desai-scratch) [inline src/client/src/shell/routes/ShellRoute.tsx]: You should be able to:

```import {ExternalWindow} from 'rt-components'```

File: src/client/src/shell/routes/ShellRoute.tsx [hunk 3]
```diff
@@ -75,8 +94,8 @@ class ShellRoute extends PureComponent<Props> {
 }
 
 const mapStateToProps = (state: GlobalState) => ({
-  displayBlotter: displayBlotterSelector(state),
-  displayAnalytics: displayAnalyticsSelector(state),
+  blotter: blotterSelector(state),
```
Comment (bhavesh-desai-scratch) [inline src/client/src/shell/routes/ShellRoute.tsx]: Maybe use hooks.

File: src/client/src/ui/drag/drag.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+export const withDrag = () => {
+  let dragNode: HTMLElement | undefined
+  let dragX = 0
+  let dragY = 0
+
+  const createDragImage = function($node: HTMLElement) {
+    const clonedNode = $node.cloneNode(true) as HTMLElement
+    clonedNode.style.position = 'absolute'
+
+    clonedNode.style.width = $node.offsetWidth + 'px'
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/drag/drag.ts]: String template please

File: src/client/src/ui/drag/drag.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+export const tilesAreDraggabe = !Environment.isRunningInIE()
+
+export const withDrag = () => {
+  let dragNode: HTMLElement | undefined
+  let dragX = 0
+  let dragY = 0
+
+  const createDragImage = function($node: HTMLElement) {
+    const clonedNode = $node.cloneNode(true) as HTMLElement
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/drag/drag.ts]: Could you add a few comments here, please?

File: src/client/src/ui/drag/drag.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import { Environment } from 'rt-system'
+
+export const tilesAreDraggabe = !Environment.isRunningInIE()
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/drag/drag.ts]: Might be worth checking again

File: src/client/src/ui/drag/drag.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    //required for FF as its drag events have no XY coords attached
+    document.ondragover = function(dragEvent) {
+      dragX = dragEvent.pageX
+      dragY = dragEvent.pageY
+    }
+
+    const dt = event.dataTransfer
+    const element = event.currentTarget
+    if (typeof dt.setDragImage === 'function') {
+      setTimeout(function() {
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/drag/drag.ts]: lambda

File: src/client/src/ui/drag/drag.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      const div = document.createElement('div')
+      div.style.display = 'none'
+      dt.setDragImage(div, 0, 0)
+    }
+  }
+
+  const onDrag = (event: React.DragEvent<HTMLDivElement>) => {
+    event.stopPropagation()
+    if (dragNode) {
+      dragNode.style.left = Math.max(0, dragX) + 'px'
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/drag/drag.ts]: Might be worth doing this in a RAF

Comment (bhavesh-desai-scratch) [reply src/client/src/ui/drag/drag.ts]: requestAnimationFrame(()=>{
      dragNode.style.left = Math.max(0, dragX) + 'px'
      dragNode.style.top = Math.max(0, dragY) + 'px'
})

File: src/client/src/ui/workspace/selectors.ts [hunk 9]
```diff
@@ -9,13 +10,18 @@ export interface ExternalWindowProps {
   browserConfig: { center: Center }
 }
 
-const makeExternalWindowProps: (key: string) => ExternalWindowProps = (key: string) => ({
+const makeExternalWindowProps: (
+  key: string,
+  tileLayout?: TileLayoutState,
+) => ExternalWindowProps = (key: string, tileLayout?: TileLayoutState) => ({
```
Comment (bhavesh-desai-scratch) [inline src/client/src/ui/workspace/selectors.ts]: Should this be optional.

File: src/client/src/shell/layouts/reducer.ts [hunk 10]
```diff
@@ -1,19 +1,29 @@
 import { LAYOUT_ACTION_TYPES, LayoutActions } from './layoutActions'
 import { externalWindowDefault } from '../../rt-components'
 
+export interface TileLayoutState {
```
Comment (bhavesh-desai-scratch) [inline src/client/src/shell/layouts/reducer.ts]: Leta change the name. WindowPosition

File: src/client/package.json [hunk 11]
```diff
@@ -116,7 +116,7 @@
     "@types/react-copy-to-clipboard": "^4.2.6",
     "@types/react-dom": "^16.8.2",
     "@types/react-helmet": "^5.0.7",
-    "@types/react-redux": "^6.0.0",
+    "@types/react-redux": "^7.1.0",
```
[the remark this answers is not in the document]
Comment (stewart-adaptive) [reply src/client/package.json]: I get errors if I don't include the types

**Pull request merged**